### DISCUSSION TITLE
How to Properly Configure Android Aot Publish

### DISCUSSION THREAD

**Discussion by vitash** [AvaloniaUI/Avalonia#17738] [Q&A] [answered]
I'm having all sorts of problems with the project I generated through the template, adding the aot option:
``` xml
    <AndroidEnableProfiledAot>True</AndroidEnableProfiledAot>
    <UseFastDeployment>false</UseFastDeployment>
    <RunAOTCompilation>true</RunAOTCompilation>
```
- Added these three options to `csproj`, The packaged apk is **180Mb**.Unacceptable, I'd expect a simple demo apk to be around 50Mb, no more than 100Mb max.
![image](https://github.com/user-attachments/assets/4466f995-9a8c-4792-946a-e6e36d4c9b4c)

- Continue adding `<PublishTrimmed>true</PublishTrimmed>`, The packaged apk is **51Mb**. Trying to run the program, it immediately exits.
- Trying to add the `<PublishAot>true</PublishAot>` option gets you an error: `Cross-OS native compilation is not supported`.
---
Is it possible to add an example of an Android Aot to `AvaloniaUI.QuickGuides`?
In talking to my network groupmates, I realized that other people's non-aot apk sizes are around 70Mb, while the apk I used for the template demo has 140Mb?

**Comment by maxkatz6** [comment, marked as answer]
> Continue adding <PublishTrimmed>true</PublishTrimmed>, The packaged apk is 51Mb. Trying to run the program, it immediately exits.

You should go from there.
You need `PublishTrimmed` to be enabled, if you want smaller binaries.
When trimming is enabled, .NET compiler will remove all the code it dimed unused (which is not always the case). Any reflection usage, for example, is potentially dangerous.
Typical problem can be with using default `ViewLocator.cs`. To make it more trimming friendly, consider replacing Activator calls with switch case for your viewmodels/views, or removing it completely.

Also, see logs and debug output, they usually help here. Android diagnostics is more complicated that desktop, as .NET ecosystem is primarily desktop friendly. So, you can enable PublishTrimmed and test your desktop project (even if you don't want to target desktop, it's still easier to debug this way). Once they work - you can continue with android.

> Is it possible to add an example of an Android Aot to AvaloniaUI.QuickGuides?

I am not aware if these quick guides has android projects connected to them. cc @jmacato @timunie

**Reply by maxkatz6** [reply]
> <RunAOTCompilation>true</RunAOTCompilation>

Typically, with Mono Android on .NET, AOT compilation increases binary size, but speeds up app startup. Just FYI.

**Reply by vitash** [reply]
Thank you for your answer, the first time to develop android really encountered a lot of problems.

**Reply by timunie** [reply]
I could maybe add a TodoList advanced to Avalonia.Samples which target all platforms and is AOT compatible. But I'd need help testing iOS. 

Unlikely to happen this year 😉
